### PR TITLE
Add Security Section

### PR DESCRIPTION
**Pull request by npm1**
The Privacy Section will be rewritten in a followup. This PR adds a section for Security considerations

***
<a href="https://pr-preview.s3.amazonaws.com/npm1/FedCM/pull/273.html" title="Last updated on Jun 14, 2022, 7:53 PM UTC (7b0937f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/fedidcg/FedCM/273/7da41fa...npm1:7b0937f.html" title="Last updated on Jun 14, 2022, 7:53 PM UTC (7b0937f)">Diff</a>

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -1518,7 +1518,66 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+
+* The token and revocation fetches expose the most: they are peformed with cookies, client id, and
+    referrer. The former is gated on the user interacting with the user agent UI, and enables the
+    [=IDP=] to communicate to the [=RP=] the information required to perform a federated
+    signin/signup. It is not possible for the [=RP=] or the [=IDP=] to force this fetch to happen,
+    as the user agent cannot be spoofed or otherwise tricked. The later can only be performed once a
+    token has been created, but does not require user interaction. That said, performing this fetch
+    does not provide any new information to the [=RP=] or the [=IDP=]. Its purpose is only to make
+    bookkeeping easier, as at this point both the [=RP=] and the [=IDP=] have information about the
+    user. The revocation obviously requires collaboration from both parties: a malicious [=RP=]
```
[the remark this answers is not in the document]
Comment (samuelgoto) [reply spec/index.bs:1654]: > I think that makes sense generally for parts other than network fetches

Something that is still not connecting to me is that the network fetches are mitigations for Privacy concerns, *not* Security, right? That is, which requests are made with cookies and which aren't, and why, that fits into the Privacy section, not the security section, right?

File: spec/index.bs [hunk 2]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
 
-Note: go over security.
+This section provides a few of the security considerations for the FedCM API. Note that there is a
+separate section for [[#privacy]] considerations.
+
+<!-- ============================================================ -->
+## Network requests ## {#network-requests}
+<!-- ============================================================ -->
+
+The FedCM API introduces the ability for a site to ask the browser to execute a few different
```
Comment (samuelgoto) [inline spec/index.bs:1600]: This is a really solid section, which I think we should keep, but I think this is misplaced: we didn't break the network requests in the specific way you are documenting here for **security** reasons, but rather we did it for **privacy** reasons, meaning that this is well tied as a mitigation to one of the **privacy** concerns. 

So, I'd suggest moving this (in this PR, or in another PR, if you want to re-do the privacy section separately).

Does that make sense?

File: spec/index.bs [hunk 3]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+│ token_endpoint           │  yes    │  yes      │  yes     │
+├──────────────────────────┼─────────┼───────────┼──────────┤
+│ client_metadata_endpoint │  no     │  yes      │  yes     │
+├──────────────────────────┼─────────┼───────────┼──────────┤
+│ revocation_endpoint      │  yes    │  yes      │  yes     │
+└──────────────────────────┴─────────┴───────────┴──────────┘
+</script>
+
+For fetches that are sent with cookies, we actually consider them equivalent to first-party fetches,
+and hence include
```
Comment (samuelgoto) [inline spec/index.bs:1622]:
```suggestion
and hence include first-party cookies as if the resource was loaded as a first-party, e.g. regardless of the SameSite value (which is used when a resource loaded as a third-party, not first-party).
```

File: spec/index.bs [hunk 4]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+[SameSite](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-the-samesite-attribute-2)=None
+cookies. This makes it easy for an [=IDP=] to adopt the FedCM API without introducing security
+issues on the API, since the [=RP=] cannot inspect the results from the fetches in any way.
+
+* The manifest fetch is safe because it is performed without cookies, [=client id=], or referrer.
```
Comment (samuelgoto) [inline spec/index.bs:1627]:
```suggestion
* The manifest fetch can't be used to track users because it is performed without cookies, [=client id=], or referrer.
```

File: spec/index.bs [hunk 5]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    Thus, anyone could perform this fetch, and the information contained therein is considered
+    public.
+
+* The accounts are fetched with cookies from the [=IDP=]. This in theory is a new power that the
+    [=RP=] gains that it would not have otherwise. Preventing too many of these fetches may be important,
+    but [=IDP=]s are already expected to protect against DoS attacks. In addition, the user agent
+    should only allow one FedCM flow per page at any given moment, immediately rejecting any
+    attempts to start another one. Since a FedCM flow can only be terminated when the user interacts
+    or after a long timer, the number of fetches performed is not a concern. The [=IDP=] does learn
+    a lot about the user from this fetch, but this is discussed in detail in the [[#privacy]]
```
Comment (samuelgoto) [inline spec/index.bs:1637]:
```suggestion
    a lot about the website (e.g. in the form of a [Referer](link to referer spec) header or the [=client_id=] value) the user is in from this fetch, and this is discussed in detail in the [[#privacy]]
```

File: spec/index.bs [hunk 6]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    section.
+
+* The client metadata fetch is performed without cookies from the [=IDP=] but with a [=client id=]
```
Comment (samuelgoto) [inline spec/index.bs:1640]:
```suggestion
* The client metadata fetch can't be used to track users too because it is performed without cookies from the [=IDP=], albeit with a [=client id=]
```

File: spec/index.bs [hunk 7]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    and a referrer. This allows the [=IDP=] to communicate the relevant [=Privacy Policy=] and
+    [=Terms of Service=] to the user agent, in case they need to be displayed. Again, the [=RP=]
```
Comment (samuelgoto) [inline spec/index.bs:1642]:
```suggestion
    [=Terms of Service=] to the user agent, in case they need to be displayed. Again, besides possible timing attacks described here, the [=RP=]
```

File: spec/index.bs [hunk 8]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    gains nothing from this fetch, and the [=RP=] could already perform this fetch if it wanted to
+    since it involves no cookies from the [=IDP=].
+
+* The token and revocation fetches expose the most: they are peformed with cookies, [=client id=],
```
Comment (samuelgoto) [inline spec/index.bs:1646]:
```suggestion
* By design, the token and revocation fetches expose the user at the website to the [=IDP=]: they are peformed with cookies, [=client id=],
```

File: spec/index.bs [hunk 9]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    and referrer. The former is gated on the user interacting with the user agent UI, and enables
+    the [=IDP=] to communicate to the [=RP=] the information required to perform a federated
+    signin/signup. It is not possible for the [=RP=] or the [=IDP=] to force this fetch to happen,
```
Comment (samuelgoto) [inline spec/index.bs:1649]: > It is not possible for the [=RP=] or the [=IDP=] to force this fetch to happen,

Can you expand on this a bit more? It is hard to comprehend what you are trying to convey here.

File: spec/index.bs [hunk 10]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+    bookkeeping easier, as at this point both the [=RP=] and the [=IDP=] have information about the
+    user. The revocation obviously requires collaboration from both parties: a malicious [=RP=]
+    cannot sign the user out of their accounts on other [=RP=]s.
+
+<!-- ============================================================ -->
+## Content Security Policy ## {#content-security-policy}
+<!-- ============================================================ -->
+
+The first fetch triggered by the FedCM API is the manifest list, which is public. However, this one
+points to the actual internal manifest, so it is important to perform a [[!CSP]] check on this to
```
Comment (samuelgoto) [inline spec/index.bs:1662]: > check on this to prevent some sneaky script from introducing FedCM calls into the [RP](https://pr-preview.s3.amazonaws.com/npm1/FedCM/pull/273.html#rp)'s site

I don't think I understand what is the threat vector here. Can you expand this a bit more on what specific problem we are trying to solve here? What's the attack here?

File: spec/index.bs [hunk 11]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+The non-same-origin fetches include, for example, the brand icon. The user agent does not perform a
+[[!CSP]] check on these because they are directly specified from the manifest. In addition, the
+rendering of this image is performed by the user agent, and as such this image cannot affect the
+[=RP=] site nor can they be inspected by the [=RP=] in any way.
+
+<!-- ============================================================ -->
+## Sec-FedCM-CSRF ## {#sec-fedcm-csrf}
+<!-- ============================================================ -->
+
+The FedCM API introduces a bunch of endpoints on the [=IDP=], so these need to be protected from XSS
```
Comment (samuelgoto) [inline spec/index.bs:1676]:
```suggestion
The FedCM API introduces a many non-static endpoints on the [=IDP=], so these need to be protected from XSS
```

File: spec/index.bs [hunk 12]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+attacks. In order to do so, the FedCM API introduces the [[#Sec-FedCM-CSRF]] header, a [=forbidden
+header name=]. An [=IDP=] is encouraged to check for this header in the requests it receives, which
```
Comment (samuelgoto) [inline spec/index.bs:1678]:
```suggestion
header name=] which cannot be set by random websites (so that the [=IDP=] can be confident that the request was originated by the FedCM browser rather than sent by a websites trying to run an XSS). An [=IDP=] is encouraged to check for this header in the requests it receives, which
```

Comment (samuelgoto) [inline spec/index.bs:1678]:
```suggestion
header name=]. An [=IDP=] must check for this header in the id_token and revocation endpoints, which
```

Comment (samuelgoto) [reply spec/index.bs:1678]: This isn't something that they should check for non-credentialed fetches, right (e.g. the manifest)?

File: spec/index.bs [hunk 13]
```diff
@@ -1585,7 +1590,110 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+ensures that the request was initiated by the user agent, based on the FedCM API. A malicious actor
+cannot spam FedCM API calls, so this is sufficient protection for the new [=IDP=] endpoints.
+
+<!-- ============================================================ -->
+## UI faking ## {#ui-faking}
+<!-- ============================================================ -->
+
+The FedCM API introduces new user agent UI, and the user agent may choose to show the UI entirely
+on top of the page's contents, if anything because the page was the one responsible for this UI.
+This introduces a potential concern of a malicious site to try to replicate the FedCM UI. This is
```
Comment (samuelgoto) [inline spec/index.bs:1688]: Could you write this in terms of a threat vector? what specifically are we trying to prevent here? What would allow a website to replicate the FedCM that would be harmful?

File: spec/index.bs (hunk 9 above)
Comment (npm1) [reply spec/index.bs]: Will reword to `It is not possible for the [=RP=] or the [=IDP=] to force the token fetch to happen without user consent`

File: spec/index.bs (hunk 10 above)
Comment (npm1) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 11 above)
Comment (npm1) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 12 above)
Comment (npm1) [reply spec/index.bs]: Done

Comment (npm1) [reply spec/index.bs]: Done

File: spec/index.bs (hunk 13 above)
Comment (npm1) [reply spec/index.bs]: Aren't all those questions are answered in this paragraph? Essentially 1) it cannot be replicated 2) because no sensitive information is requested from the user, any harmful such fake would have to be different.

File: spec/index.bs [hunk 14]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+This section provides a few of the security considerations for the FedCM API. Note that there is a
+separate section for [[#privacy]] considerations.
+
+<!-- ============================================================ -->
+## Content Security Policy ## {#content-security-policy}
+<!-- ============================================================ -->
+
+The first fetch triggered by the FedCM API is the manifest list, which is public. However, this one
+points to the actual internal manifest, so it is important to perform a [[!CSP]] check on this. This
+prevents a malicious script from introducing a FedCM prompt that is not desired by the [=RP=] or the
```
Comment (samuelgoto) [inline spec/index.bs:1602]: I'm not sure if I understand the threat model here, but wouldn't it be possible for an attacker to point to an arbitrary IDP that isn't in the CSP list?

Regardless, could you expand more on what specific threat we are trying to resolve here? You say "this prevents a malicious script from introducing a FedCM prompt", but I'm not sure what you mean by this? Is the malicious script running in the RP? In the IDP? Both? Is the attack against the IDP? The RP? What does this have to do with the manifest fetch?

And then, what precisely does the CSP help with?

File: spec/index.bs [hunk 15]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+The FedCM API introduces several non-static endpoints on the [=IDP=], so these need to be protected
+from XSS attacks. In order to do so, the FedCM API introduces the [[#Sec-FedCM-CSRF]] header, a
+[=forbidden header name=]. The header cannot be set by random websites, so the [=IDP=] can be
+confident that the request was originated by the FedCM browser rather than sent by a websites trying
+to run an XSS attack. An [=IDP=] must to check for this header in the credentialed requests it receives,
+which ensures that the request was initiated by the user agent, based on the FedCM API. A malicious
+actor cannot spam FedCM API calls, so this is sufficient protection for the new [=IDP=] endpoints.
+
+<!-- ============================================================ -->
+## UI faking ## {#ui-faking}
```
Comment (samuelgoto) [inline spec/index.bs:1626]:
```suggestion
## Browser Surface Impersonation ## {#browser-surface-impersonation}
```

File: spec/index.bs [hunk 16]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+<!-- ============================================================ -->
+
+The FedCM API introduces new user agent UI, and the user agent may choose to show the UI entirely
```
Comment (samuelgoto) [inline spec/index.bs:1629]:
```suggestion
The FedCM API introduces new (trusted) user agent UI, and the user agent may choose to show the UI entirely
```

File: spec/index.bs [hunk 17]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+on top of the page's contents, if anything because the page was the one responsible for this UI.
+This introduces a potential concern of a malicious site to try to replicate the FedCM UI. This is
```
Comment (samuelgoto) [inline spec/index.bs:1631]:
```suggestion
This introduces a potential concern of a malicious site to try to replicate the FedCM UI, gain the user's trust that the user is interacting with a trusted browser surface and gather information from the user that they would only give to the browser rather than the site (e.g. usernames/passwords of **another** site). This is
```

File: spec/index.bs [hunk 18]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+not possible because the FedCM UI requires knowledge of the user accounts within the [=IDP=]. If
```
Comment (samuelgoto) [inline spec/index.bs:1632]:
```suggestion
not possible because the FedCM UI uses the metadata about the user accounts of the [=IDP=], which the website doesn't have access. If
```

File: spec/index.bs [hunk 19]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+this is a malicious site, it would not know the user accounts, otherwise the user is already
+compromised. In addition, the FedCM UI does not prompt the user to provide additional information
+such as username or password. Thus, UI attempting to retrieve sensitive information from the user
```
Comment (samuelgoto) [inline spec/index.bs:1635]:
```suggestion
such as username or password. Thus, the website UI (e.g. iframes) attempting to retrieve sensitive information from the user
```

File: spec/index.bs (hunk 2 above)
Comment (samuelgoto) [reply spec/index.bs:1649]: LGTM

File: spec/index.bs [hunk 20]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+<!-- ============================================================ -->
+
+The FedCM API introduces the ability for a site to ask the browser to execute a few different
+network requests, as shown in the [[#high-level-design]] section. It is important for the browser to
+execute these in such a way that it does not accidentally leak sensitive information about the user
+to the site using FedCM. The following table has information about the network requests performed:
+
+<script type="text/typogram">
+┌──────────────────────────┬─────────┬───────────┬──────────┐
+│ Endpoint                 │ Cookies │ Client_id │ Referrer │
```
Comment (samuelgoto) [inline spec/index.bs:1656]:
```suggestion
│ Endpoint                 │ cookies │ client_id │ referrer │
```

File: spec/index.bs [hunk 21]
```diff
@@ -1585,12 +1590,122 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+[SameSite](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#name-the-samesite-attribute-2)
+value (which is used when a resource loaded as a third-party, not first-party). This makes it easy
+for an [=IDP=] to adopt the FedCM API without introducing security issues on the API, since the
+[=RP=] cannot inspect the results from the fetches in any way.
+
+* The manifest fetch can't be used to track users because it is performed without cookies, [=client
+    id=], or referrer. Thus, anyone could perform this fetch, and the information contained therein
+    is considered public.
+
+* The accounts are fetched with cookies from the [=IDP=]. This in theory is a new power that the
```
Comment (samuelgoto) [inline spec/index.bs:1682]:
```suggestion
* The accounts fetch can't be used to track users because it is performed with cookies from the [=IDP=] but, importantly, without the [=client id=] or referrer. This in theory is a new power that the
```

**Review by samuelgoto — COMMENTED** [review]
Looks generally good to me. Final round of review, will merge once resolved.

Biggest concern is with the description of the CSP feature, which is unclear to me what purpose it serve (I'm sure there is one, I just can't figure that out by reading this text, so please expand it a bit more to explain what specific problem does it solve).

File: spec/index.bs (hunk 14 above)
Comment (npm1) [reply spec/index.bs]: Sorry for not being clear here. I tried to rephrase this to make it clearer.

**Comment by npm1** [comment]
Ping

File: spec/index.bs [hunk 22]
```diff
@@ -1414,8 +1419,8 @@ The <dfn>fetch the internal manifest</dfn> algorithm accepts a {{FederatedIdenti
     1. Return the result of fetching the |manifest url| and parsing it into a [=Manifest=].
         TODO: specify how the fetching and the parsing works.
         The fetch MUST be done with the [[#Sec-FedCM-CSRF|Sec-FedCM-CSRF]] header but
-	without the [=Identity Provider=]'s cookies. This request MUST NOT follow
-        [[RFC7231#header.location|HTTP redirects]] and instead abort with an
+	      without the [=Identity Provider=]'s cookies. This request MUST NOT follow
+        [[RFC9110#header.location|HTTP redirects]] and instead abort with an
```
Comment (samuelgoto) [inline spec/index.bs:1423]: In a follow up PR, could you explain why we disable following redirects in the security (or privacy?) section? If easy in this PR, add a TODO so we don't forget?

File: spec/index.bs [hunk 23]
```diff
@@ -1585,12 +1590,129 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+<!-- ============================================================ -->
+## Content Security Policy ## {#content-security-policy}
+<!-- ============================================================ -->
+
+The first fetches triggered by the FedCM API are the manifest list, which is public, and the
+internal manifest. Imagine a malicious script attempting to execute the FedCM API calls to a
+malicious [=IDP=], one which is not trusted by the [=RP=]. If the call is successful, this would
+introduce browser UI on the [=RP=] with sign in options into a malicious [=IDP=]. This malicious
+[=IDP=] could then attempt to trick the user. The protection against this attack is the [[!CSP]]
+check, which would fail because the manifest would not be an origin allowed by the [=RP=], hence
```
Comment (samuelgoto) [inline spec/index.bs:1605]:
```suggestion
check, which would fail because the origin of the manifest of the malicious [=IDP=] would not be an origin included in the allowlist specified by the [[!CSP]] of the [=RP=], hence
```

File: spec/index.bs [hunk 24]
```diff
@@ -1585,12 +1590,129 @@ Note: go over how we are planning to deal with backwards compatibility.
...
-Note: go over security.
+This section provides a few of the security considerations for the FedCM API. Note that there is a
+separate section for [[#privacy]] considerations.
+
+<!-- ============================================================ -->
+## Content Security Policy ## {#content-security-policy}
+<!-- ============================================================ -->
+
+The first fetches triggered by the FedCM API are the manifest list, which is public, and the
+internal manifest. Imagine a malicious script attempting to execute the FedCM API calls to a
```
Comment (samuelgoto) [inline spec/index.bs:1601]:
```suggestion
internal manifest. Imagine a malicious script included by (and running as) the [=RP=] attempting to execute the FedCM API calls to a
```

File: spec/index.bs [hunk 25]
```diff
@@ -1585,12 +1590,129 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+
+The FedCM API introduces new (trusted) user agent UI, and the user agent may choose to show the UI
+entirely on top of the page's contents, if anything because the page was the one responsible for
+this UI. This introduces a potential concern of a malicious site to try to replicate the FedCM UI,
+gain the user's trust that the user is interacting with a trusted browser surface, and gather
+information from the user that they would only give to the browser rather than the site (e.g.
+usernames/passwords of **another** site). This is not possible because the FedCM UI uses the
+metadata about the user accounts of the [=IDP=], which the website doesn't have access. If this is a
+malicious site, it would not know the user accounts, otherwise the user is already compromised. In
+addition, the FedCM UI does not prompt the user to provide additional information such as username
```
Comment (samuelgoto) [inline spec/index.bs:1640]:
```suggestion
addition, the FedCM UI deliberately constructed to not prompt the user to provide additional information, such as username
```

File: spec/index.bs [hunk 26]
```diff
@@ -1585,12 +1590,129 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+or password. Thus, the website UI (e.g. iframes) attempting to retrieve sensitive information from
```
Comment (samuelgoto) [inline spec/index.bs:1641]:
```suggestion
or password. Thus, an attacker trying to impersonate the browser using exclusively UI that is accessible to the content area (e.g. iframes) to attempt to retrieve sensitive information from
```

File: spec/index.bs [hunk 27]
```diff
@@ -1585,12 +1590,129 @@ Note: go over how we are planning to deal with backwards compatibility.
...
+├──────────────────────────┼─────────┼───────────┼──────────┤
+│ token_endpoint           │  yes    │  yes      │  yes     │
+├──────────────────────────┼─────────┼───────────┼──────────┤
+│ client_metadata_endpoint │  no     │  yes      │  yes     │
+├──────────────────────────┼─────────┼───────────┼──────────┤
+│ revocation_endpoint      │  yes    │  yes      │  yes     │
+└──────────────────────────┴─────────┴───────────┴──────────┘
+</script>
+
+For fetches that are sent with cookies, we actually consider them equivalent to first-party fetches,
```
Comment (samuelgoto) [inline spec/index.bs:1676]: Add a TODO here to cover the reason why we don't follow the redirects?

**Review by samuelgoto — APPROVED** [review]
LGTM

Will merge once last round of comments are addressed.

Thanks @npm1 !

File: spec/index.bs (hunk 22 above)
Comment (npm1) [reply spec/index.bs:1423]: Added TODO.

**Comment by samuelgoto** [comment]
Thanks! Merging!
